### PR TITLE
test: optimise test with chai custom method overwrite

### PR DESCRIPTION
**Pull request by MaikoTan**

### CODE REVIEW

File: __tests__/index.spec.ts [hunk 1]
```diff
@@ -1,30 +1,48 @@
...
+  }
+}
+
+function noop() {
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  return function () {}
+}
+
+Assertion.overwriteChainableMethod('a', a, noop)
+Assertion.overwriteChainableMethod('an', a, noop)
```
Comment (SaarChaffee) [inline __tests__/index.spec.ts]:
```suggestion
import { Assertion, expect, util } from 'chai'

import { defaultRules, getNews, LodestoneNews, regions, News } from '../src'

const UTC_0_2022_07_19_07_59_59 = 1658217599 * 1000 // 2022-07-19 07:59:59 UTC+0

Assertion.addProperty('newsList', function () {
  new Assertion(this._obj).to.be.instanceof(Array<News>)
  this._obj.forEach(n => {
    new Assertion(n).to.have.deep.property('title').which.is.a('string')
    new Assertion(n).to.have.deep.property('epoch').which.is.a('number')
    new Assertion(n).to.have.deep.property('url').which.is.a('string')
    new Assertion(n).to.have.deep.property('date').which.is.a('date')
  })
  this.assert(
    this._obj instanceof Array<News>,
    'expected #{this} to be an array of News',
    'expected #{this} to not be an array of News',
    null,
  )
})

Assertion.addMethod('haveLen', function (n?: number) {
  if (util.flag(this, 'newsList')) {
    new Assertion(this._obj).to.have.deep.property('length').which.is.a('number')
    if (n) {
      new Assertion(this._obj).to.have.lengthOf(n)
      this.assert(
        this._obj.length === n,
        'expected #{this} to have #{exp} news, but got #{act}',
        'expected #{this} to not have #{exp} news, but got #{act}',
        n,
        this._obj.length,
      )
    } else {
      new Assertion(this._obj).to.have.lengthOf.greaterThan(0)
      this.assert(
        this._obj.length > 0,
        'expected #{this} to have more than one news, but got #{exp}',
        'expected #{this} to not have more than one news, but got #{exp}',
        this._obj.length,
      )
    }
  }
})
```

File: __tests__/index.spec.ts [hunk 2]
```diff
@@ -1,30 +1,48 @@
...
+    categories.forEach((category) => {
+      it(`should fetch ${category}`, async () => {
+        const news = await getNews({
+          region,
+          category,
+          count: 20,
+        })
+        expect(news).is.an('array')
+        expect(news).has.length.greaterThan(0)
+        expect(news[0]).is.a('news')
```
Comment (SaarChaffee) [inline __tests__/index.spec.ts]: 现在可以这么用，`haveLen()` 入参为空则判断大于0
```suggestion
        expect(news).is.a.newsList.that.haveLen()
```

File: __tests__/index.spec.ts [hunk 3]
```diff
@@ -83,9 +95,6 @@ it('should filter with after', async () => {
     after: new Date(UTC_0_2022_07_19_07_59_59),
   })
   expect(news).is.an('array')
-  expect(news[0]).has.property('title').which.is.a('string')
-  expect(news[0]).has.property('epoch').which.is.a('number')
-  expect(news[0]).has.property('url').which.is.a('string')
-  expect(news[0]).has.property('date').which.is.a('date')
-  expect(news[news.length - 1].date).is.greaterThan(new Date(UTC_0_2022_07_19_07_59_59))
+  expect(news[0]).is.a('news')
```
Comment (SaarChaffee) [inline __tests__/index.spec.ts]:
```suggestion
  expect(news).is.a.newsList.that.haveLen()
```

File: __tests__/index.spec.ts [hunk 4]
```diff
@@ -51,10 +69,7 @@ it('should accept custom config', async () => {
   const news = await lodestone.getNews({ category: 'custom' })
   expect(news).is.an('array')
   expect(news.length).equals(10)
-  expect(news[0]).has.property('title').which.is.a('string')
-  expect(news[0]).has.property('epoch').which.is.a('number')
-  expect(news[0]).has.property('url').which.is.a('string')
-  expect(news[0]).has.property('date').which.is.a('date')
+  expect(news[0]).is.a('news')
```
Comment (SaarChaffee) [inline __tests__/index.spec.ts]:
```suggestion
  expect(news).is.a.newsList.that.haveLen(10)
```

File: __tests__/index.spec.ts [hunk 5]
```diff
@@ -66,11 +81,8 @@ it('should filter with after / before', async () => {
     before: new Date('2022-07-20'),
   })
   expect(news).is.an('array')
-  expect(news[0]).has.property('title').which.is.a('string')
-  expect(news[0]).has.property('epoch').which.is.a('number')
-  expect(news[0]).has.property('url').which.is.a('string')
-  expect(news[0]).has.property('date').which.is.a('date')
-  expect(news.length).is.equal(2)
+  expect(news[0]).to.be.a('news')
+  expect(news).has.length(2)
```
Comment (SaarChaffee) [inline __tests__/index.spec.ts]:
```suggestion
  expect(news).is.a.newsList.that.haveLen(2)
```

**Review by SaarChaffee — CHANGES_REQUESTED** [review]

File: __tests__/index.spec.ts [hunk 6]
```diff
@@ -29,12 +40,10 @@ Assertion.overwriteChainableMethod('an', a, noop)
...
-    new Assertion(n).to.have.property('title').which.is.a('string')
-    new Assertion(n).to.have.property('epoch').which.is.a('number')
-    new Assertion(n).to.have.property('url').which.is.a('string')
-    new Assertion(n).to.have.property('date').which.is.a('date')
-  })
+  new Assertion(this._obj).all.have.property('title').which.is.a('string')
+  new Assertion(this._obj).all.have.property('epoch').which.is.a('number')
+  new Assertion(this._obj).all.have.property('url').which.is.a('string')
+  new Assertion(this._obj).all.have.property('date').which.is.a('date')
 })
```
Comment (SaarChaffee) [inline __tests__/index.spec.ts]:
```suggestion
  util.flag(this, 'newsList', true)
})
```

**Review by SaarChaffee — CHANGES_REQUESTED** [review]

File: __tests__/index.spec.ts [hunk 7]
```diff
@@ -27,6 +27,40 @@ function noop() {
...
+
+function length(_super: typeof Assertion) {
+  return function (this: typeof Assertion, n?: number, message?: string) {
+    if (util.flag(this, 'newsList')) {
+      new Assertion(this._obj).to.have.property('length').which.is.a('number')
+      if (n) {
+        new Assertion(this._obj).to.have.length(n, message)
+      } else {
+        new Assertion(this._obj).to.have.length.greaterThan(0, message)
+      }
```
Comment (SaarChaffee) [inline __tests__/index.spec.ts]:
```suggestion
      }
      util.flag(this, 'newsList', false)
```

**Review by SaarChaffee — CHANGES_REQUESTED** [review]

**Review by SaarChaffee — APPROVED** [review]

**Comment by MaikoTan** [comment]
[The lodestone is maintaining currently](https://jp.finalfantasyxiv.com/lodestone/news/detail/5bcbc0d889444667265d97d324205537b508973b), which broke the test always.

**Comment by SaarChaffee** [comment]
```log
 29 passing (3m)
 1 failing
```

**Comment by MaikoTan** [comment]
you can trigger the ci test by clicking the re-run button.

File: __tests__/index.spec.ts [hunk 8]
```diff
@@ -49,12 +99,7 @@ it('should accept custom config', async () => {
     },
   })
   const news = await lodestone.getNews({ category: 'custom' })
-  expect(news).is.an('array')
-  expect(news.length).equals(10)
-  expect(news[0]).has.property('title').which.is.a('string')
-  expect(news[0]).has.property('epoch').which.is.a('number')
-  expect(news[0]).has.property('url').which.is.a('string')
-  expect(news[0]).has.property('date').which.is.a('date')
+  expect(news).to.be.newsList.which.has.length(20)
```
Comment (SaarChaffee) [inline __tests__/index.spec.ts]:
```suggestion
  expect(news).to.be.newsList.which.has.length(10)
```

**Review by SaarChaffee — CHANGES_REQUESTED** [review]

**Review by SaarChaffee — APPROVED** [review]

**Pull request merged**